javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -246,6 +246,34 @@ void setCurrentArray() {
         };
     }
 
+    @SuppressWarnings("unchecked")
+    <T2> void visit(LeafVisitor<T2[]> visitor) {
+        int start = lastDigit(offset);
+        for (int index = 0; index < length; ) {
+            final T2[] leaf = (T2[]) getLeaf(index);
+            final int end = Math.min(leaf.length, start + length - index);
+
+            visitor.visit(leaf, start, end);
+
+            index += end - start;
+            start = 0;
+        }
+    }
+
+    <U> BitMappedTrie<U> map(Function<? super T, ? extends U> mapper) {
```

Comment:
I also think we do not need the further improvement (for now).

With the new visitor, the `map` method looks like this:

```java
    <U> BitMappedTrie<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper, "mapper is null");
        final Object[] results = new Object[length()];
        this.<T> visit((leaf, start, end, index) -> {
            for (int i = start; i < end; i++) {
                elems[index++] = mapper.apply(leaf[i]);
            }
            return index;
        });
        return BitMappedTrie.ofAll(elems, length);
    }
```